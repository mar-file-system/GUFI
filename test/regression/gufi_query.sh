#!/usr/bin/env bash

# This file is part of GUFI, which is part of MarFS, which is released
# under the BSD license.
#
#
# Copyright (c) 2017, Los Alamos National Security (LANS), LLC
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without modification,
# are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this
# list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation and/or
# other materials provided with the distribution.
#
# 3. Neither the name of the copyright holder nor the names of its contributors
# may be used to endorse or promote products derived from this software without
# specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
# IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
# INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
# OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
# ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
#
# From Los Alamos National Security, LLC:
# LA-CC-15-039
#
# Copyright (c) 2017, Los Alamos National Security, LLC All rights reserved.
# Copyright 2017. Los Alamos National Security, LLC. This software was produced
# under U.S. Government contract DE-AC52-06NA25396 for Los Alamos National
# Laboratory (LANL), which is operated by Los Alamos National Security, LLC for
# the U.S. Department of Energy. The U.S. Government has rights to use,
# reproduce, and distribute this software.  NEITHER THE GOVERNMENT NOR LOS
# ALAMOS NATIONAL SECURITY, LLC MAKES ANY WARRANTY, EXPRESS OR IMPLIED, OR
# ASSUMES ANY LIABILITY FOR THE USE OF THIS SOFTWARE.  If software is
# modified to produce derivative works, such modified software should be
# clearly marked, so as not to confuse it with the version available from
# LANL.
#
# THIS SOFTWARE IS PROVIDED BY LOS ALAMOS NATIONAL SECURITY, LLC AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
# THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL LOS ALAMOS NATIONAL SECURITY, LLC OR
# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
# OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
# OF SUCH DAMAGE.



set -e

ROOT="$(realpath ${BASH_SOURCE[0]})"
ROOT="$(dirname ${ROOT})"
ROOT="$(dirname ${ROOT})"
ROOT="$(dirname ${ROOT})"

GUFI_QUERY="${ROOT}/src/gufi_query"

# output directories
SRCDIR="prefix"
INDEXROOT="${SRCDIR}.gufi"

source ${ROOT}/test/regression/setup.sh "${ROOT}" "${SRCDIR}" "${INDEXROOT}"

OUTPUT="gufi_query.out"

function replace() {
    echo "$@" | sed "s/[[:space:]]*$//g; s/${GUFI_QUERY//\//\\/}/gufi_query/g; s/${INDEXROOT//\//\\/}\\///g; s/\\/${SRCDIR//\//\\/}/./g"
}

(
export LC_ALL=C

echo "# Get only directories"
replace "$ ${GUFI_QUERY} -d \" \" -S \"SELECT name FROM summary\" ${INDEXROOT}"
output=$(${GUFI_QUERY} -d " " -S "SELECT name FROM summary" ${INDEXROOT} | sort)
replace "${output}"
echo

echo "# Get only non-directories"
replace "$ ${GUFI_QUERY} -d \" \" -E "SELECT name FROM entries" ${INDEXROOT}"
output=$(${GUFI_QUERY} -d " " -E "SELECT name FROM entries" ${INDEXROOT} | sort)
replace "${output}"
echo

echo "# Get all directory and non-directory names only"
replace "$ ${GUFI_QUERY} -d \" \" -S \"SELECT name FROM summary\" -E \"SELECT name FROM entries\" ${INDEXROOT}"
output=$(${GUFI_QUERY} -d " " -S "SELECT name FROM summary" -E "SELECT name FROM entries" ${INDEXROOT} | sort)
replace "${output}"
echo

echo "# Get relative paths of all directories and non-directories ascending names"
replace "$ ${GUFI_QUERY} -d \" \" -e 0 -a -I \"CREATE TABLE out(name TEXT)\" -S \"INSERT INTO out SELECT path(name) FROM summary\" -E \"INSERT INTO out SELECT path((SELECT name FROM summary WHERE summary.inode == pentries.pinode)) || '/' || name FROM pentries\" -J \"INSERT INTO aggregate.out SELECT * FROM out\" -G \"SELECT name FROM out ORDER BY name ASC\" ${INDEXROOT}"
output=$(${GUFI_QUERY} -d " " -e 0 -a -I "CREATE TABLE out(name TEXT)" -S "INSERT INTO out SELECT path(name) FROM summary" -E "INSERT INTO out SELECT path((SELECT name FROM summary WHERE summary.inode == pentries.pinode)) || '/' || name FROM pentries" -J "INSERT INTO aggregate.out SELECT * FROM out" -G "SELECT name FROM out ORDER BY name ASC" ${INDEXROOT})
replace "${output}" | sed "s/${INDEXROOT//\//\\/}/./g"
echo

echo "# Get relative paths of all directories and non-directories descending names"
replace "$ ${GUFI_QUERY} -d \" \" -e 0 -a -I \"CREATE TABLE out(name TEXT)\" -S \"INSERT INTO out SELECT path(name) FROM summary\" -E \"INSERT INTO out SELECT path((SELECT name FROM summary WHERE summary.inode == pentries.pinode)) || '/' || name || '/' || name FROM pentries\" -J \"INSERT INTO aggregate.out SELECT * FROM out\" -G \"SELECT name FROM out ORDER BY name DESC\" ${INDEXROOT}"
output=$(${GUFI_QUERY} -d " " -e 0 -a -I "CREATE TABLE out(name TEXT)" -S "INSERT INTO out SELECT path(name) FROM summary" -E "INSERT INTO out SELECT path((SELECT name FROM summary WHERE summary.inode == pentries.pinode)) || '/' || name FROM pentries" -J "INSERT INTO aggregate.out SELECT * FROM out" -G "SELECT name FROM out ORDER BY name DESC" ${INDEXROOT})
replace "${output}" | sed "s/${INDEXROOT//\//\\/}/./g"
echo

echo "# Get relative paths of all directories and non-directories ascending sizes"
replace "$ ${GUFI_QUERY} -d \" \" -e 0 -a -I \"CREATE TABLE out(name TEXT, size INT64)\" -E \"INSERT INTO out SELECT path(name) || '/' || name, size FROM pentries\" -J \"INSERT INTO aggregate.out SELECT * FROM out\" -G \"SELECT name FROM out ORDER BY size ASC, name ASC\" ${INDEXROOT}"
output=$(${GUFI_QUERY} -d " " -e 0 -a -I "CREATE TABLE out(name TEXT, size INT64)" -E "INSERT INTO out SELECT path((SELECT name FROM summary WHERE summary.inode == pentries.pinode)) || '/' || name, size FROM pentries" -J "INSERT INTO aggregate.out SELECT * FROM out" -G "SELECT name FROM out ORDER BY size ASC, name ASC" ${INDEXROOT})
replace "${output}"
echo

echo "# Get relative paths of all directories and non-directories descending sizes"
replace "$ ${GUFI_QUERY} -d \" \" -e 0 -a -I \"CREATE TABLE out(name TEXT, size INT64)\" -E \"INSERT INTO out SELECT path((SELECT name FROM summary WHERE summary.inode == pentries.pinode)) || '/' || name, size FROM pentries\" -J \"INSERT INTO aggregate.out SELECT * FROM out\" -G \"SELECT name FROM out ORDER BY size DESC, name DESC\" ${INDEXROOT}"
output=$(${GUFI_QUERY} -d " " -e 0 -a -I "CREATE TABLE out(name TEXT, size INT64)" -E "INSERT INTO out SELECT path((SELECT name FROM summary WHERE summary.inode == pentries.pinode)) || '/' || name, size FROM pentries" -J "INSERT INTO aggregate.out SELECT * FROM out" -G "SELECT name FROM out ORDER BY size DESC, name DESC" ${INDEXROOT})
replace "${output}"
echo
) | tee "${OUTPUT}"

diff ${ROOT}/test/regression/gufi_query.expected "${OUTPUT}"
rm "${OUTPUT}"
