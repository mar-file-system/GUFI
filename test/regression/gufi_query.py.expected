# gufi_query.py help
$ gufi_query.py -h
usage: gufi_query [options] GUFI_tree ...
options:
  -h                     help
  -H                     show assigned input values (debugging)
  -v                     version
  -T <SQL_tsum>          SQL for tree-summary table
  -S <SQL_sum>           SQL for summary table
  -E <SQL_ent>           SQL for entries table
  -a <0|1|2>             0 - if returned row, run next SQL, else stop (continue descent) (default)
                         1 - skip T, run S and E whether or not a row was returned (old -a)
                         2 - run T, S, and E whether or not a row was returned
  -n <threads>           number of threads
  -j                     print the information in terse form
  -o <out_fname>         output file (one-per-thread, with thread-id suffix)
  -d <delim>             delimiter (one char)  [use 'x' for 0x1E]
  -O <out_DB>            output DB
  -u                     prefix row with 1 int column count and each column with 1 octet type and 1 size_t length
  -I <SQL_init>          SQL init
  -F <SQL_fin>           SQL cleanup
  -y <min level>         minimum level to go down
  -z <max level>         maximum level to go down
  -J <SQL_interm>        SQL for intermediate results
  -K <create aggregate>  SQL to create the final aggregation table
  -G <SQL_aggregate>     SQL for aggregated results
  -m                     Keep mtime and atime same on the database files
  -B <buffer size>       size of each thread's output buffer in bytes
  -w                     open the database files in read-write mode instead of read only mode
  -x                     index/query xattrs
  -k <filename>          file containing directory names to skip
  -M <bytes>             target memory footprint
  -s <path>              File name prefix for swap files
  -p <path>              Source path prefix for %s in SQL
  -e                     compress work items
  -Q <basename>
     <table>
     <template>.<table>
     <view>              External database file basename, per-attach table name, template + table name, and the resultant view
  -D <filename>          File containing paths at single level to index (not including starting path). Must also use -y

GUFI_tree        find GUFI tree here


# Walk without returning results
$ gufi_query.py

# Get only directories
$ gufi_query.py -d " " -n 2 -S "SELECT name FROM vrsummary;"
directory
empty_directory
leaf_directory
prefix
subdirectory
unusual#? directory ,

# Get only non-directories
$ gufi_query.py -d " " -n 2 -E "SELECT name FROM vrpentries;"
.hidden
1KB
1MB
directory_symlink
executable
file_symlink
leaf_file1
leaf_file2
old_file
readonly
repeat_name
repeat_name
unusual, name?#
writable

# Get all directory and non-directory names
$ gufi_query.py -d " " -n 2 -S "SELECT name FROM vrsummary;" -E "SELECT name FROM vrpentries;"
.hidden
1KB
1MB
directory
directory_symlink
empty_directory
executable
file_symlink
leaf_directory
leaf_file1
leaf_file2
old_file
prefix
readonly
repeat_name
repeat_name
subdirectory
unusual#? directory ,
unusual, name?#
writable

